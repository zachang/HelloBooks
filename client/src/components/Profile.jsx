import React from 'react';
import classnames from 'classnames';
import PropTypes from 'react-proptypes';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { getOneUserAction, updateUserAction } from '../actions/userAction';

/**
 * Profile class declaration
 * @class Profile
 * @extends {React.Component}
 */
export class Profile extends React.Component {
  /**
   * class constructor
   * @param {object} props
   */
  constructor(props) {
    super(props);
    this.state = {
      userData: {
        fullname: '',
        username: '',
        email: '',
        phone_no: '',
        user_image: '',
        user_image_txt: ''
      },
      errors: null,
      showToast: false,
      imagePreviewUrl: '',
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleImageChange = this.handleImageChange.bind(this);
  }

  /**
   * @method componentDidMount
   * @return {void} void
   */
  componentDidMount() {
    this.props.getOneUserAction(this.props.params.id);
  }

  /**
   * @method componentWillReceiveProps
   * @param {object} nextProps - nextProps
   * @return {object} nextProps
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.userState.success === true) {
      if (this.state.showToast) {
        Materialize.toast('User updated!', 4000);
        this.setState({
          showToast: false
        });
      }
      this.setState({
        userData: nextProps.userState.user
      });
    } else if (nextProps.userState.errors === 'Error, No update done') {
      if (this.state.showToast) {
        Materialize.toast('User not updated!', 4000);
        this.setState({ showToast: false });
      }
      this.setState({
        errors: nextProps.userState.errors
      });
    } else if (!nextProps.userState.success && nextProps.userState.errors !== 'Error, No update done') {
      if (this.state.showToast) {
        Materialize.toast('Fill the form properly!', 4000);
        this.setState({ showToast: false });
      }
      this.setState({
        errors: nextProps.userState.errors
      });
    }
  }

  /**
   * Handles user Image file input
   * @method handleImageChange
   * @return {void} void
   * @param {object} event - event
   */
  handleImageChange(event) {
    const reader = new FileReader();
    const userData = this.state.userData;
    const file = event.target.files[0];
    userData.user_image = file;
    userData.user_image_text = file.name;

    reader.onloadend = () => {
      this.setState({
        userData,
        imagePreviewUrl: reader.result
      });
    };

    reader.readAsDataURL(userData.user_image);
  }

  /**
   * Handles user update input
   * @method handleChange
   * @return {void} void
   * @param {object} event - event
   */
  handleChange(event) {
    const userData = this.state.userData;
    userData[event.target.name] = event.target.value;
    this.setState({
      userData,
      errors: null
    });
  }

  /**
   * Handles user form update
   * @method handleSubmit
   * @return {void}
   * @param {object} event - event
   */
  handleSubmit(event) {
    event.preventDefault();
    this.setState({
      showToast: true
    });
    this.props.updateUserAction(this.state.userData, this.props.params.id);
  }
  /**
   * Renders Profile component
   * @return {XML} JSX
   */
  render() {
    const { userData, imagePreviewUrl  } = this.state;

    return (
      <div>
        {userData && <div>
          <div className='row'>
            <div className='row'>
              <div className='col s10 m8 l6 profile_marg' style={{ marginLeft: '23%', marginTop: '7%' }}>
                <div className='card' style={{boxShadow: '2px 1px 7px #000'}}>
                  <form method='post' className='col s12' onSubmit={this.handleSubmit}>
                    <div className='card-content white-text'
                      style={{ height: '200px', background: 'rgba(77, 182, 172, 0.9)' }}
                    >

                      <div className='profile-image circle'>
                        {(imagePreviewUrl) ? <img
                          className='activator image-display circle'
                          src={imagePreviewUrl}/>
                            :
                          <img
                          className='activator image-display circle'
                          src={(userData.user_image !== null)
                          ? userData.user_image : '../../imgs/avatar.png'}
                        />}
                      </div>

                    </div>
                    <div className='card-action' style={{ height: '340px', background: 'rgba(217,217,217,0.7)' }}>

                      <div className='row'>
                        <div className='input-field col s6'>
                          <input
                            id='fullname'
                            name='fullname'
                            type='text'
                            className={classnames({
                              invalid: (this.state.errors && !!this.state.errors.fullname) ?
                                !!this.state.errors.fullname : false
                            })}
                            value={userData.fullname || ''}
                            onChange={this.handleChange}
                          />
                          <label
                            htmlFor='fullname'
                            className={((this.state.errors && !!this.state.errors.fullname)
                              || userData.fullname.length > 0) ? 'custom-active custom-validate' : 'custom-validate'}
                            data-error={(this.state.errors && !!this.state.errors.fullname) ?
                              this.state.errors.fullname : ''}
                          >
                            Full Name
                          </label>
                        </div>
                        <div className='input-field col s6'>
                          <input
                            id='username'
                            name='username'
                            type='text'
                            className={classnames({
                              invalid: (this.state.errors && !!this.state.errors.username) ?
                                !!this.state.errors.username : false
                            })}
                            value={userData.username || ''}
                            onChange={this.handleChange}
                          />
                          <label
                            htmlFor='username'
                            className={((this.state.errors && !!this.state.errors.username)
                              || userData.username.length > 0) ? 'custom-active custom-validate' : 'custom-validate'}
                            data-error={(this.state.errors && !!this.state.errors.username) ?
                              this.state.errors.username : ''}
                          >
                            Username
                          </label>
                        </div>
                      </div>

                      <div className='row'>
                        <div className='input-field col s6'>
                          <input
                            id='email'
                            name='email'
                            type='text'
                            className={classnames({
                              invalid: (this.state.errors && !!this.state.errors.email) ?
                                !!this.state.errors.email : false
                            })}
                            value={userData.email || ''}
                            onChange={this.handleChange}
                          />
                          <label
                            htmlFor='email'
                            className={((this.state.errors && !!this.state.errors.email)
                              || userData.email.length > 0) ? 'custom-active custom-validate' : 'custom-validate'}
                            data-error={(this.state.errors && !!this.state.errors.email) ?
                              this.state.errors.email : ''}
                          >
                            Email
                          </label>
                        </div>
                        <div className='input-field col s6'>
                          <input
                            id='phone_no'
                            name='phone_no'
                            type='text'
                            className={classnames({
                              invalid: (this.state.errors && !!this.state.errors.phone_no) ?
                                !!this.state.errors.phone_no : false
                            })}
                            value={userData.phone_no || ''}
                            onChange={this.handleChange}
                          />
                          <label
                            htmlFor='phone number'
                            className={((this.state.errors && !!this.state.errors.phone_no)
                              || userData.phone_no.length > 0) ? 'custom-active custom-validate' : 'custom-validate'}
                            data-error={(this.state.errors && !!this.state.errors.phone_no) ?
                              this.state.errors.phone_no : ''}
                          >
                            Phone Number
                          </label>
                        </div>
                      </div>

                      <div className='row'>
                        <div className='file-field input-field col s10'>
                          <div className='btn'>
                            <span>Add Image</span>
                            <input
                              type='file'
                              id='user_image'
                              name='user_image'
                              accept='image/*'
                              onChange={this.handleImageChange}
                            />
                          </div>
                          <div className='file-path-wrapper'>
                            <input
                              type='text'
                              name='user_image_txt'
                              className={classnames('file-path', {
                                invalid: (this.state.errors && !!this.state.errors.user_image_txt) ?
                                  !!this.state.errors.user_image_txt : false
                              })}
                            />
                          </div>
                        </div>
                      </div>

                      <button
                        className='waves-effect waves-light btn profile-btn'
                        style={{marginLeft: '35%'}}
                        type='submit'
                      >
                        Update
                      </button>

                    </div>
                  </form>
                </div>
              </div>
            </div>

          </div>
        </div>}
      </div>

    );
  }
}

Profile.propTypes = {
  userState: PropTypes.object.isRequired,
  getOneUserAction: PropTypes.func.isRequired,
  updateUserAction: PropTypes.func.isRequired
};
const mapStateToProps = state => ({
  userState: state.userReducer,
});
const mapDispatchToProps = dispatch =>
  bindActionCreators({ getOneUserAction, updateUserAction }, dispatch);

export default connect(mapStateToProps, mapDispatchToProps)(Profile);