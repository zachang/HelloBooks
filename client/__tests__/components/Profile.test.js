import React from 'react';
import toJson from 'enzyme-to-json';
import sinon from 'sinon';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import storeConfiguration from '../../src/store/store';
import { Profile } from '../../src/components/Profile.jsx';


const event = {
  preventDefault: jest.fn(),
  target: {
    value: 'mmmmmm',
    files: ['eben.jpg']
  }
};

const blob = new Blob(['Baked.jpg'], { type: 'image/jpeg' });
blob.name = "file.png";
const eventImage = {
  preventDefault: jest.fn(),
  target: {
    files: [blob]
  }
};

const userData = {
  fullname: '',
  username: '',
  email: '',
  phoneNo: '',
  userImage: '',
  userImage_txt: ''
};

describe('<Profile/>', () => {
  // create a spy function for getOneUserAction
  const getOneUserAction = sinon.spy();
  // create a spy function for getCategoryAction
  const updateUserAction = sinon.spy();
  // spy on handleImageChange of Profile Page
  sinon.spy(Profile.prototype, 'handleImageChange');
  // spy on handleSubmit of Profile Page
  sinon.spy(Profile.prototype, 'handleSubmit');
  // spy on handleChange of Profile Page
  sinon.spy(Profile.prototype, 'handleChange');
  // spy on componentWillMount of Profile Page
  sinon.spy(Profile.prototype, 'componentWillMount');
  // spy on componentWillReceiveProps of Profile Page
  sinon.spy(Profile.prototype, 'componentWillReceiveProps');

  const showToast = true;
  const props = {
    userState: {
      success: false,
      errors: null,
      fails: null,
      users: [],
      user: [],
      pageCount: null,
    },
    showToast,
    getOneUserAction,
    updateUserAction
  };

  const middleware = [thunk];
  const mockStore = configureMockStore(middleware);
  let store;
  let shallowComponent;
  beforeEach(() => {
    store = mockStore(storeConfiguration);
    shallowComponent = shallow(<Profile {...props} />);
  });

  it('renders <Profile /> component', () => {
    const tree = toJson(shallowComponent);
    expect(tree).toMatchSnapshot();
  });

  it('should check that handleSubmit is called', () => {
    shallowComponent.instance().handleSubmit(event);
    expect(Profile.prototype.handleSubmit.called).toEqual(true);
  });

  it('should check that handleImageChange is called', () => {
    shallowComponent.instance().handleImageChange(eventImage);
    expect(Profile.prototype.handleImageChange.called).toEqual(true);
  });

  it('should check that handleChange is called', () => {
    shallowComponent.instance().handleChange(event);
    expect(Profile.prototype.handleChange.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when profile update', () => {
    props.userState.success = true;
    shallowComponent.setState({
      errors: {
        fullname: 'kkk',
        username: 'kkkk',
        email: 'mmm',
        phoneNo: 'mmm',
        userImage: 'mmm',
        userImage_txt: 'mmmm'
      },
      userData: {
        fullname: 'mmmmmmmm',
        username: '',
        email: '',
        phoneNo: '',
        userImage: '',
        userImage_txt: ''
      },
      imagePreviewUrl: '',
      showToast: true
    });
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(Profile.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when error on profile update', () => {
    props.userState.success = false;
    props.userState.errors = 'Error, No update done';
    shallowComponent.setState({
      errors: {
        fullname: 'kkk',
        username: 'kkkk',
        email: 'mmm',
        phoneNo: 'mmm',
        userImage: 'mmm',
        userImage_txt: 'mmmm'
      },
      userData: {
        fullname: 'mmmmmmmm',
        username: '',
        email: '',
        phoneNo: '',
        userImage: '',
        userImage_txt: ''
      },
      imagePreviewUrl: '',
      showToast: true
    });
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(Profile.prototype.componentWillReceiveProps.called).toEqual(true);
  });

  it('calls componentWillReceiveProps when error on profile update', () => {
    props.userState.success = false;
    props.userState.errors = 'Error';
    shallowComponent.setState({
      errors: {
        fullname: 'kkk',
        username: 'kkkk',
        email: 'mmm',
        phoneNo: 'mmm',
        userImage: 'mmm',
        userImage_txt: 'mmmm'
      },
      userData: {
        fullname: 'mmmmmmmm',
        username: '',
        email: '',
        phoneNo: '',
        userImage: '',
        userImage_txt: ''
      },
      imagePreviewUrl: '',
      showToast: true
    });
    shallowComponent.instance().componentWillReceiveProps(props);
    expect(Profile.prototype.componentWillReceiveProps.called).toEqual(true);
  });
});