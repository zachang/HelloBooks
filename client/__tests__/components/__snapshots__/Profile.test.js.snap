// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Profile/> renders <Profile /> component 1`] = `
<div>
  <div>
    <div
      className="row"
    >
      <div
        className="row"
      >
        <div
          className="col s10 m8 l6 profile_marg"
          style={
            Object {
              "marginLeft": "23%",
              "marginTop": "7%",
            }
          }
        >
          <div
            className="card"
            style={
              Object {
                "boxShadow": "2px 1px 7px #000",
              }
            }
          >
            <form
              className="col s12"
              method="post"
              onSubmit={[Function]}
            >
              <div
                className="card-content white-text"
                style={
                  Object {
                    "background": "rgba(77, 182, 172, 0.9)",
                    "height": "200px",
                  }
                }
              >
                <div
                  className="profile-image circle"
                >
                  <img
                    className="activator image-display circle"
                    src=""
                  />
                </div>
              </div>
              <div
                className="card-action"
                style={
                  Object {
                    "background": "rgba(217,217,217,0.7)",
                    "height": "340px",
                  }
                }
              >
                <div
                  className="row"
                >
                  <div
                    className="input-field col s6"
                  >
                    <input
                      className=""
                      id="fullname"
                      name="fullname"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                    <label
                      className="custom-validate"
                      data-error=""
                      htmlFor="fullname"
                    >
                      Full Name
                    </label>
                  </div>
                  <div
                    className="input-field col s6"
                  >
                    <input
                      className=""
                      id="username"
                      name="username"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                    <label
                      className="custom-validate"
                      data-error=""
                      htmlFor="username"
                    >
                      Username
                    </label>
                  </div>
                </div>
                <div
                  className="row"
                >
                  <div
                    className="input-field col s6"
                  >
                    <input
                      className=""
                      id="email"
                      name="email"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                    <label
                      className="custom-validate"
                      data-error=""
                      htmlFor="email"
                    >
                      Email
                    </label>
                  </div>
                  <div
                    className="input-field col s6"
                  >
                    <input
                      className=""
                      id="phoneNo"
                      name="phoneNo"
                      onChange={[Function]}
                      type="text"
                      value=""
                    />
                    <label
                      className="custom-validate"
                      data-error=""
                      htmlFor="phone number"
                    >
                      Phone Number
                    </label>
                  </div>
                </div>
                <div
                  className="row"
                >
                  <div
                    className="file-field input-field col s10"
                  >
                    <div
                      className="btn"
                    >
                      <span>
                        Add Image
                      </span>
                      <input
                        accept="image/*"
                        id="userImage"
                        name="userImage"
                        onChange={[Function]}
                        type="file"
                      />
                    </div>
                    <div
                      className="file-path-wrapper"
                    >
                      <input
                        className="file-path"
                        name="userImage_txt"
                        type="text"
                      />
                    </div>
                  </div>
                </div>
                <button
                  className="waves-effect waves-light btn profile-btn"
                  style={
                    Object {
                      "marginLeft": "35%",
                    }
                  }
                  type="submit"
                >
                  Update
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
